### PR TITLE
[RFC] Update boost to 1.5.7

### PR DESCRIPTION
**Pull request by Paxxi**
I started looking into this to silence the annoying unkown compiler spam that we have on win32 currently

uc

The simple option would be to disable boost compiler config with a define but I assume that could lead to less optimal code.

I grabbed the latest boost and used their bcp tool to only grab the parts we're using multi_index and circular_buffer and their dependencies. This resulted in a much smaller package than what we currently have. Around 2k files and <2mb compressed for both win32 and unified depends. Compare this to current unified depends package at 40mb

Another positive effect of this is that we can remove some of our boost patches that are included and we can completely remove the building of boost, we only use the headers.

Code in this PR is ugly and a quick hack to start a discussion.
the archives can be found at
http://delusional.nu/boost_1_57_0-win32.7z and http://delusional.nu/boost_1_57_0.tar.bz2

### CODE REVIEW

**Comment by Paxxi** [comment]
@wsnipex @Karlson2k @MartijnKaijser what do you think? it will obviously require some testing but so far it's looking good

**Comment by Memphiz** [comment]
I would say it now installs nothing anymore for unified depends - not even the headers - the whole makefile looks like a NOP to me now...

**Comment by Karlson2k** [comment]
I already did the same for win32 during crash fixing.
https://github.com/Karlson2k/xbmc/commit/2400d571c2f86c58f23696c22bc63666e6c126d3
It was tested alongside with other fixes in test builds.

**Comment by wsnipex** [comment]
yeah, that Makefile won't fly ;) If we really only need the headers, you'll have to at least copy them to $(PREFIX)/include

**Comment by Paxxi** [comment]
@karlson2k no reported issues?

**Comment by Karlson2k** [comment]
@Paxxi My version for win32 is fine.
For POSIX I'd recommend to name archive to something like `boost_1_57_0-reduced.tar.bz2` or `stripped`. And share your preparation script, so other devs will be able to reproduce and update lib.

**Comment by Paxxi** [comment]
@Karlson2k but your boost lacks the multi_index stuff from what I can see?
Anyway this was to see the response before committing any real time to it. I'm taking the lack of screams of agony as a good sign and will do this properly

**Comment by Karlson2k** [comment]
@Paxxi It's good idea and I vote for it. Boost has too many headers so even unpacking boost on jenkins takes noticeable time for each build.
Just make it reproducible and share the script/receipt. It should be solidrock stable reproducible, verified two or three times that result is the same from the scratch, not a single accidental build. :)

**Comment by notspiff** [comment]
my .001c;

macine time is cheap, bandwidth is cheap, human time is expensive. why waste resources on saving some machine time? you'd have to redo this every time you bump boost, or at least make sure it works.

**Comment by Paxxi** [comment]
Fair point @notspiff but the process takes about 5-10min with compile times and our current boost is from 2011 so I don't forsee this taking up a lot of time

**Comment by Paxxi** [comment]
@Karlson2k what do you think of the instructions? they're in script format and should be copy/pastable
http://delusional.nu/boost_1_57_0_stripped.tar.bz2 and http://delusional.nu/boost_1_57_0_stripped-win32.7z

**Comment by Paxxi** [comment]
@Memphiz and @wsnipex (when back) does it look more acceptable now?

**Comment by FernetMenta** [comment]
can't we just get rid of this boost crap completely?

**Comment by Paxxi** [comment]
Sadly not, our current font cache is based on boost multiindex, montellese uses the sha1 and guid parts, probably some more that's still using it

**Comment by FernetMenta** [comment]
thanks for clarification

**Comment by Paxxi** [comment]
jenkins bubild this please

**Comment by Paxxi** [comment]
jenkins build this please

**Comment by stefansaraev** [comment]
@Paxxi http://mirrors.kodi.tv/build-deps/sources/boost_1_57_0_stripped.tar.bz2 does not exist, perhaps you forgot to upload ?

**Comment by Paxxi** [comment]
jenkins build this please

File: tools/depends/target/boost/Makefile [hunk 1]
```diff
@@ -15,23 +15,15 @@ $(TARBALLS_LOCATION)/$(ARCHIVE):
...
-	echo '	<cflags>"$(CFLAGS)"' >> $(PLATFORM)/tools/build/v2/user-config.jam
-	echo '  <cxxflags>"$(CXXFLAGS)"' >> $(PLATFORM)/tools/build/v2/user-config.jam
-	echo '	;' >> $(PLATFORM)/tools/build/v2/user-config.jam
-	cd $(PLATFORM); patch -p1 < ../add-arm-mem-barrier.patch
-	cd $(PLATFORM); patch -p1 < ../fix-deprecated-swp.patch
 ifeq (arm, $(findstring arm, $(CPU)))
-	cd $(PLATFORM); patch -p0 < ../fix_shared_ptr_multithread.patch
+	#ugly hack to avoid issue with line endings. boost headers use \r\n and patch
+	#will be converted by git to \n and fail to apply
+	cd $(PLATFORM); cat ../fix_shared_ptr_multithread.patch >> boost/config/user.hpp
```
Comment (Paxxi) [inline tools/depends/target/boost/Makefile]: @Memphiz @wsnipex I welcome any suggestions on how to solve this in prettier way.

**Comment by Paxxi** [comment]
hopefully fixed build issue
prettier commit message
jenkins build this please again

File: tools/depends/target/boost/Makefile (hunk 1 above)
Comment (Memphiz) [inline tools/depends/target/boost/Makefile]: Does this give any help?

http://stackoverflow.com/questions/7754972/how-do-i-get-patch-to-ignore-carriage-returns

Comment (Paxxi) [inline tools/depends/target/boost/Makefile]: Thanks that might work but I think ill leave it as is with echo >>

File: tools/depends/target/boost/Makefile [hunk 2]
```diff
@@ -15,23 +15,15 @@ $(TARBALLS_LOCATION)/$(ARCHIVE):
...
-	cd $(PLATFORM); patch -p1 < ../fix-deprecated-swp.patch
 ifeq (arm, $(findstring arm, $(CPU)))
-	cd $(PLATFORM); patch -p0 < ../fix_shared_ptr_multithread.patch
+	#ugly hack to avoid issue with line endings. boost headers use \r\n and patch
+	#will be converted by git to \n and fail to apply
+	cd $(PLATFORM); echo "#define BOOST_SP_USE_SPINLOCK" >> boost/config/user.hpp
 endif
-	cd $(PLATFORM); patch -p1 < ../fix-self-assignment-warnings.patch
-	cd $(PLATFORM); ./bootstrap.sh --prefix=$(PREFIX)
+	cd $(PLATFORM); cp -r boost $(PREFIX)/include/
```
Comment (koying) [inline tools/depends/target/boost/Makefile]: Why a plain "cp"?
bootstrap/bjam is still the recommended install way: http://theboostcpplibraries.com/introduction-installation

Comment (koying) [inline tools/depends/target/boost/Makefile]: Ah, because it's a "redux" package? Pff...

**Comment by koying** [comment]
I'll -1 this one, for common depends anyway.
This is far too much hacking for just removing win32 warnings on a sensible lib that we want gone.

If you want the warnings gone, please limit the PR to win32.

**Comment by Paxxi** [comment]
Hacking? The stripping is done using boosts own tool, please tell me how this is in any way worse than the boost we have?

**Comment by koying** [comment]
"Hacking" was a poorly chosen word ;)

From my perspective, I objectively just see a change to a component for no benefits but with a induced risk of destabilization on platforms using common deps, which are only android and osx, afaik.

If @Memphiz gives a +1, I won't bikeshed ;)

I'd give a +1 to importing your boost subset in-tree, otoh. Benefit would be more manageability and removing of a dependency altogether.

**Comment by Memphiz** [comment]
I am undecided too now. I know i pointed out the missing makefile change for unified depends. But the problem is i lack free time for the next couple of month - so if there are issues from that i wouldn't be able to support for osx and ios.

But i like the idea of speeding up the boost install. Do you guys think using the approach from @paxxi on unified deps but without bumping the boost version would have any major risks? So we wouldn't throw away his work (but need a stripped down tarball from the older version...)

**Comment by fritsch** [comment]
I think investing time to kill the three locations it is still used would
be much better invested. We now have C++11 and lot of features can be
replaced / reimplemented more easily.

2015-07-13 16:42 GMT+02:00 Memphiz <notifications@github.com>:

> I am undecided too now. I know i pointed out the missing makefile change
> for unified depends. But the problem is i lack free time for the next
> couple of month - so if there are issues from that i wouldn't be able to
> support for osx and ios.
>
> But i like the idea of speeding up the boost install. Do you guys think
> using the approach from @paxxi <https://github.com/paxxi> on unified deps
> but without bumping the boost version would have any major risks? So we
> wouldn't throw away his work (but need a stripped down tarball from the
> older version...)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6743#issuecomment-120952913>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by Paxxi** [comment]
Replacing boost::totally_ordered got rid of the warnings so this isnt really needed on win32 anymore, I'll close this and we can rip boost out as much as possible instead.

**Pull request closed without merging**